src/async.rs

```diff
@@ -0,0 +1,201 @@
+use std::{
+    io,
+    net::{SocketAddr, ToSocketAddrs},
+    sync::atomic::{
+        AtomicI32,
+        Ordering
+    },
+    time::Duration,
+};
+
+use tokio::net::UdpSocket;
+
+use crate::{handle_response, pdu, SnmpError, SnmpResult, SnmpMessageType, SnmpPdu, Value, BUFFER_SIZE};
+
+struct AsyncRequest {
+    socket: UdpSocket,
+    timeout: Option<Duration>,
+}
+
+impl AsyncRequest {
+    async fn send_and_recv(&mut self, pdu: pdu::Buf) -> SnmpResult<Vec<u8>> {
+        use tokio::time;
+
+        self.socket
+            .send(&pdu[..])
+            .await
+            .map_err(|_| SnmpError::SendError)?;
+
+        let mut buf = [0; BUFFER_SIZE];
+
+        let fut = self.socket
+            .recv(&mut buf);
+
+        match self.timeout {
+            Some(timeout) => {
+                time::timeout(timeout, fut)
+                .await
+                .map_err(|_| SnmpError::ReceiveError)?
+            },
+            None => {
+                fut.await
+            },
+        }
+            .map_err(|_| SnmpError::ReceiveError)
+            .map(|size| buf[0..size].into())
+    }
+}
+
+/// Asynchronous SNMPv2 client.
+pub struct AsyncSession {
+    destination: SocketAddr,
+    community: Vec<u8>,
+    timeout: Option<Duration>,
+    req_id: AtomicI32,
+}
+
+impl AsyncSession {
+    pub fn new<SA, T>(
+        destination: SA,
+        community: T,
+        timeout: Option<Duration>,
+        starting_req_id: i32,
+    ) -> io::Result<Self>
+    where
+        SA: ToSocketAddrs,
+        T: AsRef<[u8]>,
+    {
+        let address = destination
+            .to_socket_addrs()?
```

Comment:
If understand correctly [ToSocketAddrs::to_socket_addrs](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html#tymethod.to_socket_addrs) is blocking.
> Note that this function may block the current thread while resolution is performed.

Is this something we'd want to avoid/handle differently?

Reply:
The `new` function is not async so it should be ok I guess. Otherwise a non-blocking async builder can be introduced.